src/rules/no_unused_vars.rs

```diff
@@ -0,0 +1,2052 @@
+      "
+import webpack from 'webpack';
+export default function webpackLoader(this: webpack.loader.LoaderContext) {}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+import execa, { Options as ExecaOptions } from 'execa';
+export function foo(options: ExecaOptions): execa {
+  options();
+}
+      ",
+    );
+  }
+
+  #[test]
+  fn no_unused_vars_ts_ok_14() {
+    assert_lint_ok::<NoUnusedVars>(
+      "
+import { Foo, Bar } from './types';
+export class Baz<F = Foo & Bar> {}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+// warning 'B' is defined but never used
+export const a: Array<{ b: B }> = [];
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+export enum FormFieldIds {
+  PHONE = 'phone',
+  EMAIL = 'email',
+}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+enum FormFieldIds {
+  PHONE = 'phone',
+  EMAIL = 'email',
+}
+interface IFoo {
+  fieldName: FormFieldIds;
+}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+enum FormFieldIds {
+  PHONE = 'phone',
+  EMAIL = 'email',
+}
+interface IFoo {
+  fieldName: FormFieldIds.EMAIL;
+}
+      ",
+    );
+  }
+
+  #[test]
+  fn no_unused_vars_ts_ok_15() {
+    assert_lint_ok::<NoUnusedVars>(
+      "
+import * as fastify from 'fastify';
+import { Server, IncomingMessage, ServerResponse } from 'http';
+const server: fastify.FastifyInstance<
+  Server,
+  IncomingMessage,
+  ServerResponse
+> = fastify({});
+server.get('/ping');
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+declare function foo();
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+declare namespace Foo {
+  function bar(line: string, index: number | null, tabSize: number): number;
+  var baz: string;
+}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+declare var Foo: {
+  new (value?: any): Object;
+  foo(): string;
+};
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+declare class Foo {
+  constructor(value?: any): Object;
+  foo(): string;
+}
+      ",
+    );
+  }
+
+  #[test]
+  fn no_unused_vars_ts_ok_16() {
+    assert_lint_ok::<NoUnusedVars>(
+      "
+import foo from 'foo';
+export interface Bar extends foo.i18n {}
+      ",
+    );
+
+    assert_lint_ok::<NoUnusedVars>(
+      "
+import foo from 'foo';
+import bar from 'foo';
+export interface Bar extends foo.i18n<bar> {}
+      ",
+    );
+
+    // TODO(kdy1): Unignore
+
+    //     assert_lint_ok::<NoUnusedVars>(
+    //       "
+    // import { TypeA } from './interface';
+    // export const a = <GenericComponent<TypeA> />;
+    //       ",
+    //     );
+
+    // TODO(kdy1): Unignore
+
+    //     assert_lint_ok::<NoUnusedVars>(
+    //       "
+    // const text = 'text';
+    // export function Foo() {
+    //   return (
+    //     <div>
+    //       <input type=\"search\" size={30} placeholder={text} />
+    //     </div>
+    //   );
+    // }
+    //       ",
+    //     );
```

Comment:
@kdy1 do you want to uncomment his before landing?

Reply:
I'm not sure because It would require lots of work. (Currently, jsx nodes are not handled by the resolver.)

Reply:
Sounds good